torch/utils/data/distributed.py

```diff
@@ -65,3 +65,112 @@ def __len__(self):
 
     def set_epoch(self, epoch):
         self.epoch = epoch
+
+class ChunkDataReader(object):
+    r"""Reads a chunk of data given a chunk index.
+
+    A chunk could be full file, section of a large file, folders, URLs, or any other
+    abstraction that allows data to be segmented roughly the same size.
+
+    As an example, chunking could be used in a scenario where a dataset is split between
+    several CSV files each containing complete records or examples. In this case,
+    each file could be a chunk. Similarly a large CSV file can be considered as a
+    collection of chunks by defining chunk boundary based on some physical size
+    (e.g. 32MB) and seeking to the first record in each chunk during reading.
+
+    Another example would be chunks as individual folders, containing multiple audio files.
+    In this scenario, each worker would have access to different folders, 
+    where binary files would be read from the filesystem.
+
+    The reading logic must be implemented inside `__call__` method
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self, idx):
+        r"""Returns `list(samples)`
+
+        If `samples` contain tensors, numpy arrays, numbers,
+        dicts or lists, the default collate function can be implicitly used.
+        Otherwise, a custom `collate_fn` must be provided to `DataLoader`
+
+        When no more data is available, `StopIteration` must be raised.
+        """
+        raise NotImplementedError
+
+class DistributedChunkSampler(Sampler):
+    r"""This sampler introduces distributed sampling without padding and dataset dependency.
+
+    This sampler is very similar to the `DistributedSampler`, however without padding and
+    the dependency on the size of the dataset. With two levels of sampling, the
```

Comment:
How is this independent of the dataset size? You're still asking for the number of chunks. I don't see why would we need this sampler instead of only having `DistributedSampler`.

Reply:
Existing `DistributedSampler` depends on `len(dataset)`, which is not available on `IterableDataset` - when dataset size is known, `MapDataset` can be used instead. `DistributedSampler` also pads the index list, which is undesired for chunking, as it would result in different workers reading the portion of dataset again.

In the first prototype, we modified `DistributedSampler` instead of creating a new sampler, but the final design wasn't good. One of the reasons was that `DistributedSampler` would need to have the existing `dataset` changed to optional and a new optional argument `num_chunks` would be needed to replace `len(dataset)`. That leads to a constructor with all parameters being optional, with the undesired behavior of an invalid instance if the user didnt specify either `dataset` or `num_chunks`. The second issue was that lots of conditionals to disable padding and switching between `len(dataset)` and `num_chunks`, not only doubling the size the class, but also changing a specialized behavior to a generic one that could confuse the user.

Reply:
This still seems to be dependent on the length of the dataset through `num_chunks`, no?
* Is `num_chunks` the length of the `IterableDataset`? If so, then the latter can be patched to have that as a length, or just changed to a `Dataset`.
* What if `DistributedSampler` simply took the `len(dataset)` as parameter directly? That's the only way the dataset is used within. (That would be BC-breaking of course, unless we do something like detecting when a dataset is passed, and save the length instead.)

Reply:
- I might be missing here, but the [original](https://github.com/pytorch/pytorch/pull/19228) idea of `IterableDataset` was to represent data 1) without direct indexing of examples and 2) without using `len(dataset)`, which can cover lots of scenarios where computing length can be expensive (database) or impossible (a stream dataset). `IterableDataset` guarantees that `DataLoader` can iterate over it for data and when it is exhausted, an exception will be raised, signaling `Dataloader` to stop. This aspect is interesting because it lets space for future extensions, such as `InfiniteDataset` that would never be able to return its length but still would be compatible with `IterableDataset`.

By following this definition, we disabled `__len__` for `ChunkDataset(IterableDataset)`:
```python
    def __len__(self):
        # `IterableDataset` classes have unknown dataset size
        raise NotImplementedError
```

In this PR, `num_chunks` means how many parts the dataset was divided into, but it doesn't mean how many examples there are in the dataset. `num_chunks` is only used by `DistributedChunkSampler` while `ChunkDataset` keeps blind regarding its length. For MNIST training dataset, for example, there could be 12 chunks (files) with 4615 samples each and a final chunk with 4620 examples.

- If breaking BC is not an issue, `len(dataset)` would definately work. The only other minor issue we would have to address is adding an optional boolean flag that disables padding to make the list of indices evenly divisible. Changing `DistributedSampler` ctor from `def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True):` to `def __init__(self, size, num_replicas=None, rank=None, shuffle=True, padding=True):` does the trick. What do you think?

Reply:
> * I might be missing here, but the [original](https://github.com/pytorch/pytorch/pull/19228) idea of `IterableDataset` was to represent data 1) without direct indexing of examples and 2) without using `len(dataset)`

That is correct. I just wanted to make sure the meaning and use of `num_chunks` was clear :)

> * If breaking BC is not an issue, `len(dataset)` would definately work. The only other minor issue we would have to address is adding an optional boolean flag that disables padding to make the list of indices evenly divisible. Changing `DistributedSampler` ctor from `def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True):` to `def __init__(self, size, num_replicas=None, rank=None, shuffle=True, padding=True):` does the trick. What do you think?

Which line would you disable in [distributed.py](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py)?

Reply:
The new flag would disable from https://github.com/pytorch/pytorch/blob/9705d60a2f7dabe0adcc13598b0e5268678f9d00/torch/utils/data/distributed.py#L53 through https://github.com/pytorch/pytorch/blob/9705d60a2f7dabe0adcc13598b0e5268678f9d00/torch/utils/data/distributed.py#L55
when the user toggles it. Distributed data loading doesnt require dataset chunks to be equally divided between workers, especially if that incurs in reading and applying heavy transformations on duplicate chunks.

Reply:
Just to be very clear here: By "distributed" do you mean multiple machines or multiple processes?

Reply:
ChunkDataset is flexible to work with any combination among single node + single process, single node + multiple processes and multiple nodes + multiple processes. The latter is the most challenging and the one we wanted to make easier. In fact, the whole idea was that users could start with single node + single process during development and when it is ready for production, with just minor changes, it would scale to the multiple nodes + multiple processes